### ISSUE TITLE
Neo4J crashes on startup. Transaction log is 0 size.

### ISSUE THREAD

**Issue by geoghegk** [neo4j/neo4j#13048]
- Neo4j version: community edition 4.4.16
- Operating system: Red Hat Enterprise Linux release 8.7 Container running in Red Hat OpenShift Container Platform
- API/Driver: Docker image using neo4j-community-4.4.16-unix.tar.gz
- **Steps to reproduce**
This problem is only evident at one particular customer site. There are many hundreds of successful installations in house and at other customer sites. Unable to reproduce anywhere else.
- Expected behavior
Neo4j container starts up and functions correctly.
- Actual behavior
We encountered an issue at a customer site, whereby Neo4j is crashing DURING STARTUP every time. It crashes at the point where it is trying to read the transaction log file:
```
'org.neo4j.dbms.database.DefaultSystemGraphInitializer@d5b599d6' was successfully initialized, but failed to start. Please see the attached cause exception "Unable to read log version and last committed tx from '/data/transactions/system/neostore.transaction.db.0'. Was only able to read -1 bytes, but was expecting 8".
```
The file is 0 size. It seems like Neo4J has failed to write anything to that file. The file (`/data/transactions/system/neostore.transaction.db.0`) exists, in the correct location, it is readable and writable, but has 0 size. There are no errors earlier in the log that would indicate an issue during write.

This problem is only evident at one particular customer site. There are many hundreds of successful installations in house and at other customer sites.

If the pod is configured to use no persistence, then the problem goes away. When persistence is configured, the problem comes back. The PV and PVC are successfully created and PVC is being written to by neo4j - so it doesn't look like a storage permissions problem - but that database transaction file is always 0 length.

We have been unable to reproduce the issue using the same storage type in house (IBM Spectrum Scale/GPFS).

My thought on this is maybe something going wrong around file locking of the transaction log. My understanding is that the version of Neo4J that we use no longer uses system calls in its locking mechanism. Are there any dependencies on the storage/underlying filesystem here that could be causing an issue with locking? Its odd that we only see this problem at one site - any only with the IBM Spectrum Scale storage.

debug log during startup:
```
--------------------------------------------------------------------------------
                                                                                            [ Library path ]                                
                                                            --------------------------------------------------------------------------------
                                                            /usr/lib/jdk-11.0.16.1+1/lib/default
                                                            /usr/lib/jdk-11.0.16.1+1/lib
                                                            /usr/lib64
                                                            /usr/lib
                                                            
                                                            --------------------------------------------------------------------------------
                                                                                         [ System properties ]                              
                                                            --------------------------------------------------------------------------------
                                                            awt.toolkit = sun.awt.X11.XToolkit
                                                            jdk.tls.rejectClientInitiatedRenegotiation = true
                                                            file.encoding.pkg = sun.io
                                                            jdk.extensions.version = 11.0.16.1
                                                            sun.jnu.encoding = ANSI_X3.4-1968
                                                            sun.arch.data.model = 64
                                                            com.ibm.vm.bitmode = 64
                                                            user.variant = 
                                                            com.ibm.oti.vm.library.version = 29
                                                            user.timezone = UTC
                                                            sun.java.launcher = SUN_STANDARD
                                                            user.country = US
                                                            sun.boot.library.path = /usr/lib/jdk-11.0.16.1+1/lib/default:/usr/lib/jdk-11.0.16.1+1/lib
                                                            sun.java.command = org.neo4j.server.CommunityEntryPoint --home-dir=/var/lib/neo4j --config-dir=/var/lib/neo4j/conf
                                                            io.netty.tryReflectionSetAccessible = true
                                                            sun.cpu.endian = little
                                                            user.home = /
                                                            user.language = en
                                                            ibm.signalhandling.rs = false
                                                            jdk.extensions.name = Extensions for OpenJDK for Eclipse OpenJ9
                                                            file.separator = /
                                                            com.ibm.util.extralibs.properties = 
                                                            sun.java.launcher.pid = 360
                                                            com.ibm.oti.configuration = scar
                                                            com.ibm.oti.shared.enabled = false
                                                            com.ibm.oti.vm.bootstrap.library.path = /usr/lib/jdk-11.0.16.1+1/lib/default:/usr/lib/jdk-11.0.16.1+1/lib
                                                            jdk.tls.ephemeralDHKeySize = 2048
                                                            com.ibm.cpu.endian = little
                                                            user.name = 1000930000
                                                            path.separator = :
                                                            jdk.nio.maxCachedBufferSize = 262144
                                                            ibm.signalhandling.sigint = true
                                                            file.encoding = UTF-8
                                                            jnidispatch.path = /tmp/jna2021555411582016347.tmp
                                                            ibm.signalhandling.sigchain = true
                                                            jna.loaded = true
                                                            com.ibm.system.agent.path = /usr/lib/jdk-11.0.16.1+1/lib
                                                            user.dir = /var/lib/neo4j
                                                            com.ibm.jcl.checkClassPath = 
                                                            ibm.system.encoding = ANSI_X3.4-1968
                                                            sun.io.unicode.encoding = UnicodeLittle
                                                            log4j2.disable.jmx = true
                                                            
                                                            --------------------------------------------------------------------------------
                                                                                  [ (IANA) TimeZone database version ]                      
                                                            --------------------------------------------------------------------------------
                                                              TimeZone version: 2022a (available for 601 zone identifiers)
                                                            
                                                            --------------------------------------------------------------------------------
                                                                                        [ Network information ]                             
                                                            --------------------------------------------------------------------------------
                                                            Interface eth0:
                                                                address: 
                                                                address: 
                                                            Interface lo:
                                                                address: 0:0:0:0:0:0:0:1%lo
                                                                address: 127.0.0.1
                                                            
                                                            --------------------------------------------------------------------------------
                                                                                     [ Native access information ]                          
                                                            --------------------------------------------------------------------------------
                                                            Native access details: Linux native access is available.
                                                            
                                                            --------------------------------------------------------------------------------
                                                                                            [ DBMS config ]                                 
                                                            --------------------------------------------------------------------------------
                                                            DBMS provided settings:
                                                            dbms.connector.bolt.enabled=true
                                                            dbms.connector.bolt.tls_level=OPTIONAL
                                                            dbms.connector.http.enabled=true
                                                            dbms.connector.https.enabled=true
                                                            dbms.default_listen_address=0.0.0.0
                                                            dbms.directories.import=/var/lib/neo4j/import
                                                            dbms.directories.logs=/logs
                                                            dbms.directories.neo4j_home=/var/lib/neo4j
                                                            dbms.jvm.additional=-XX:+UseG1GC
                                                            -XX:-OmitStackTraceInFastThrow
-XX:+AlwaysPreTouch
-XX:+UnlockExperimentalVMOptions
-XX:+TrustFinalNonStaticFields
-XX:+DisableExplicitGC
-XX:MaxInlineLevel=15
-XX:-UseBiasedLocking
-Djdk.nio.maxCachedBufferSize=262144
-Dio.netty.tryReflectionSetAccessible=true
-Djdk.tls.ephemeralDHKeySize=2048
-Djdk.tls.rejectClientInitiatedRenegotiation=true
-XX:FlightRecorderOptions=stackdepth=256
-XX:+UnlockDiagnosticVMOptions
-XX:+DebugNonSafepoints
-Dlog4j2.disable.jmx=true
dbms.logs.debug.level=DEBUG
dbms.memory.pagecache.size=512M
dbms.ssl.policy.bolt.base_directory=/var/lib/neo4j/certificates
dbms.ssl.policy.bolt.ciphers=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
dbms.ssl.policy.bolt.enabled=true
dbms.ssl.policy.bolt.private_key=/var/lib/neo4j/certificates/tls.key
dbms.ssl.policy.bolt.public_certificate=/var/lib/neo4j/certificates/tls.crt
dbms.ssl.policy.https.base_directory=/var/lib/neo4j/certificates
dbms.ssl.policy.https.ciphers=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
dbms.ssl.policy.https.enabled=true
dbms.ssl.policy.https.private_key=/var/lib/neo4j/certificates/tls.key
dbms.ssl.policy.https.public_certificate=/var/lib/neo4j/certificates/tls.crt
dbms.tx_log.rotation.retention_policy=100M size
dbms.tx_state.memory_allocation=ON_HEAP
dbms.windows_service_name=neo4j

2023-02-14 16:16:08.571+0000 INFO [o.n.s.c.SslPolicyLoader] Loaded SSL policy 'BOLT' = SslPolicy{keyCertChain=Subject: CN=.ta-neo4j.ibm-ta.svc, Issuer: CN=openshift-service-serving-signer@1663169102, ciphers=[TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256], tlsVersions=[TLSv1.2], clientAuth=OPTIONAL}
2023-02-14 16:16:08.571+0000 INFO [o.n.s.c.SslPolicyLoader] Loaded SSL policy 'HTTPS' = SslPolicy{keyCertChain=Subject: CN=.ta-neo4j.ibm-ta.svc, Issuer: CN=openshift-service-serving-signer@1663169102, ciphers=[TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256], tlsVersions=[TLSv1.2], clientAuth=OPTIONAL}
2023-02-14 16:16:08.580+0000 INFO [o.n.d.i.DefaultIdentityModule] Found ServerId on disk: ServerId{af52aea0} (af52aea0-bf28-429d-b5b9-e2725cd1455f)
2023-02-14 16:16:08.581+0000 INFO [o.n.d.i.DefaultIdentityModule] This instance is ServerId{af52aea0} (af52aea0-bf28-429d-b5b9-e2725cd1455f)
2023-02-14 16:16:08.587+0000 INFO [o.n.g.f.EditionLocksFactories] Locking implementation 'forseti' selected.
2023-02-14 16:16:10.462+0000 INFO [o.n.s.CommunityNeoWebServer] ======== Neo4j 4.4.16 ========
2023-02-14 16:16:10.483+0000 INFO [o.n.d.d.DefaultDatabaseManager] Creating 'DatabaseId{00000000[system]}'.
2023-02-14 16:16:10.723+0000 DEBUG [i.n.h.s.JdkSslContext] Default protocols (JDK): [TLSv1.3, TLSv1.2]
2023-02-14 16:16:10.723+0000 DEBUG [i.n.h.s.JdkSslContext] Default cipher suites (JDK): [TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA, TLS_RSA_WITH_AES_128_GCM_SHA256, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_256_CBC_SHA, TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384]
2023-02-14 16:16:10.849+0000 DEBUG [i.n.u.i.NativeLibraryLoader] -Dio.netty.native.workdir: /tmp (io.netty.tmpdir)
2023-02-14 16:16:10.850+0000 DEBUG [i.n.u.i.NativeLibraryLoader] -Dio.netty.native.deleteLibAfterLoading: true
2023-02-14 16:16:10.850+0000 DEBUG [i.n.u.i.NativeLibraryLoader] -Dio.netty.native.tryPatchShadedId: true
2023-02-14 16:16:10.850+0000 DEBUG [i.n.u.i.NativeLibraryLoader] -Dio.netty.native.detectNativeLibraryDuplicates: true
2023-02-14 16:16:10.864+0000 DEBUG [i.n.u.i.NativeLibraryLoader] Successfully loaded the library /tmp/libnetty_transport_native_epoll_x86_6412630054574465331921.so
2023-02-14 16:16:10.868+0000 DEBUG [i.n.u.NetUtil] -Djava.net.preferIPv4Stack: false
2023-02-14 16:16:10.868+0000 DEBUG [i.n.u.NetUtil] -Djava.net.preferIPv6Addresses: false
2023-02-14 16:16:10.870+0000 DEBUG [i.n.u.NetUtilInitializations] Loopback interface: lo (lo, 0:0:0:0:0:0:0:1%lo)
2023-02-14 16:16:10.871+0000 DEBUG [i.n.u.NetUtil] /proc/sys/net/core/somaxconn: 128
2023-02-14 16:16:10.872+0000 INFO [o.n.b.BoltServer] Bolt server loaded
2023-02-14 16:16:10.877+0000 DEBUG [i.n.c.MultithreadEventLoopGroup] -Dio.netty.eventLoopThreads: 24
2023-02-14 16:16:10.925+0000 DEBUG [o.n.b.r.s.ExecutorBoltScheduler] Initialized bolt thread pool.
2023-02-14 16:16:10.984+0000 DEBUG [o.n.k.i.c.VmPauseMonitorComponent] Starting VM pause monitor
2023-02-14 16:16:10.987+0000 INFO [o.n.d.d.DefaultDatabaseManager] Starting 'DatabaseId{00000000[system]}'.
2023-02-14 16:16:11.309+0000 INFO [o.n.k.i.s.f.RecordFormatSelector] [system/00000000] Selected RecordFormat:PageAlignedV4_3[AF4.3.0] record format from store /data/databases/system
2023-02-14 16:16:11.311+0000 INFO [o.n.k.i.s.f.RecordFormatSelector] [system/00000000] Selected format 'RecordFormat:PageAlignedV4_3[AF4.3.0]' for existing store /data/databases/system with format 'RecordFormat:PageAlignedV4_3[AF4.3.0]'
2023-02-14 16:16:11.453+0000 WARN [o.n.k.d.Database] [system/00000000] Exception occurred while starting the database. Trying to stop already started components.
java.lang.RuntimeException: org.neo4j.kernel.impl.transaction.log.entry.IncompleteLogHeaderException: Unable to read log version and last committed tx from '/data/transactions/system/neostore.transaction.db.0'. Was only able to read -1 bytes, but was expecting 8
at org.neo4j.kernel.impl.transaction.log.files.checkpoint.DetachedLogTailScanner.findLogTail(DetachedLogTailScanner.java:102) ~[neo4j-kernel-4.4.16.jar:4.4.16]
at org.neo4j.kernel.impl.transaction.log.files.checkpoint.AbstractLogTailScanner.getTailInformation(AbstractLogTailScanner.java:246) ~[neo4j-kernel-4.4.16.jar:4.4.16]
at org.neo4j.kernel.impl.transaction.log.files.checkpoint.AbstractLogTailScanner.getTailInformation(AbstractLogTailScanner.java:237) ~[neo4j-kernel-4.4.16.jar:4.4.16]
... 23 more
2023-02-14 16:16:11.459+0000 ERROR [o.n.d.d.DefaultDatabaseManager] Failed to start DatabaseId{00000000[system]}
org.neo4j.dbms.api.DatabaseManagementException: An error occurred! Unable to start DatabaseId{00000000[system]}.
at org.neo4j.dbms.database.AbstractDatabaseManager.startDatabase(AbstractDatabaseManager.java:236) ~[neo4j-4.4.16.jar:4.4.16]
... 7 more
2023-02-14 16:16:12.480+0000 INFO [o.n.g.f.DatabaseManagementServiceFactory] Shutdown started
2023-02-14 16:16:12.480+0000 DEBUG [o.n.b.r.s.ExecutorBoltScheduler] Stopping connections
2023-02-14 16:16:12.480+0000 DEBUG [o.n.b.r.s.ExecutorBoltScheduler] Connections stopped
2023-02-14 16:16:12.480+0000 DEBUG [o.n.b.r.s.ExecutorBoltScheduler] Shutting down thread pool
2023-02-14 16:16:12.481+0000 DEBUG [o.n.b.r.s.ExecutorBoltScheduler] Thread pool shut down
2023-02-14 16:16:12.481+0000 DEBUG [o.n.b.t.NettyServer] Shutting down event loop group
```

Also raised a community discussion: https://community.neo4j.com/t/neo4j-container-crashing-on-startup-transaction-database-0-size/58554 .Outcome of that discussion was to raise an issue here.

**Comment by burqen** [comment]
Hello @geoghegk 👋 Thank you for reporting. I'll have a look into this issue.

Meanwhile, you can help me by sharing the full stacktrace, if you have it. It seems to be cutoff in the log you posted and I assume it was logged in full in `debug.log`.
```
java.lang.RuntimeException: org.neo4j.kernel.impl.transaction.log.entry.IncompleteLogHeaderException: Unable to read log version and last committed tx from '/data/transactions/system/neostore.transaction.db.0'. Was only able to read -1 bytes, but was expecting 8
at org.neo4j.kernel.impl.transaction.log.files.checkpoint.DetachedLogTailScanner.findLogTail(DetachedLogTailScanner.java:102) ~[neo4j-kernel-4.4.16.jar:4.4.16]
at org.neo4j.kernel.impl.transaction.log.files.checkpoint.AbstractLogTailScanner.getTailInformation(AbstractLogTailScanner.java:246) ~[neo4j-kernel-4.4.16.jar:4.4.16]
at org.neo4j.kernel.impl.transaction.log.files.checkpoint.AbstractLogTailScanner.getTailInformation(AbstractLogTailScanner.java:237) ~[neo4j-kernel-4.4.16.jar:4.4.16]
... 23 more
```

**Comment by burqen** [comment]
And another question. If you can share the whole `debug.log`, that would also be helpful.

**Comment by burqen** [comment]
And yet another question. What `checkpoint.x` logs do you have and how large are they?

**Comment by geoghegk** [comment]
Attaching debug.log. I don't have the checkpoint logs. Will gather next time I have time with client.

[debug.log](https://github.com/neo4j/neo4j/files/10813838/debug.log)

Looking at the code, I notice that there is a native call to preallocate the space required for the transaction log (by default 250MB). On my own dev environment, I can see that as soon as the allocation occurs (and before any writes), the file size immediately jumps to 250MB(262144000 bytes) - with the file being zeroed out. Given that the file is size 0 on the clients environment - it might be that the preallocation for that file is failing? Any known issues around that area?

**Comment by burqen** [comment]
@geoghegk There are two parts of this

**Failure to start on corrupted logs**
Starting with corrupt should at least be possible with `unsupported.dbms.tx_log.fail_on_corrupted_log_files=false`. That is not the case for this scenario, at least not from my artificial reproduction. You can ofc try it out on your end but I don't have high hopes. In any case I will patch it.

**Corrupt logs**
I haven't been able to figure out how we could end up in this state. Like you say it seems like something is going on with your specific environment. The logs that you shared only cover one startup attempt after the fact that the dbms was first created. The issue that caused the transaction logs to look like they do must have happened during the creation of the dbms. If you have those logs, please share them. There is a possibility that we might see some system call failures in that.

Given that this only happens in your specific setup and you cannot reproduce it even with the same setup, there is not so much I can do to help you further than that I'm afraid.

**Comment by geoghegk** [comment]
Hi @burqen, I built some of the neo4j jars myself and added some custom detailed logging. It turns out that on this environment, the call to posix_fallocate is not succeeding. Its returning a status code of 28:

>ENOSPC: No space left on device. This error is returned when there is no free space left on the device or filesystem that the file resides on.

The storage experts are looking into this now. There is certainly space on the disk - but the fallocate file is still failing with this error. It may be a limitation with this particular storage configuration. I'll let you know when I find out more.

However, I would say that it would good if neo4j could surface the return codes for the native calls that it makes when they are not successful.  I feel like I should have been able to get better information on this issue from the debug logging. Thanks.

**Comment by burqen** [comment]
@geoghegk That's very interesting! I completely agree that that information should surface. Can you share the stack trace where you see this error?

I see that we do have some handling of those type of errors but maybe something is missing there. The preallocation of tx-log files is an optimisation that is not strictly necessary, and it's not supported on platforms, so when the preallocation  fail we log the errors instead of throwing. Do you see any log messages like that in the debug.log?

**Comment by geoghegk** [comment]
Hi @burqen 
the piece of code i added the sys outs to is:
```
    private static NativeCallResult wrapResult( NativeCall call )
    {
        System.out.println(Instant.now() + "TA-DEBUG Entering LinuxNativeAccess.wrapResult");
        try
        {
            System.out.println(Instant.now() + "TA-DEBUG Entering LinuxNativeAccess.wrapResult before call");
            TAUtil.checkTxLogFile();
            int result = call.call();
            System.out.println(Instant.now() + "TA-DEBUG Entering LinuxNativeAccess.wrapResult after call. result(success==0) " + result);
            TAUtil.checkTxLogFile();
            if ( result == SUCCESS )
            {
                return NativeCallResult.SUCCESS;
            }
            else
            {
                System.out.println(Instant.now() + "TA-DEBUG Leaving LinuxNativeAccess.wrapResult with new NativeCallResult");
                return new NativeCallResult( result, tryExtractError( result ) );
            }
        }
        catch ( LastErrorException e )
        {
            System.out.println(Instant.now() + "TA-DEBUG Leaving LinuxNativeAccess.wrapResult with exception " + e);
            return new NativeCallResult( e.getErrorCode(), e.getMessage() );
        }
    }
```

to give:
[neo4j_extra_debug.log](https://github.com/neo4j/neo4j/files/10850545/neo4j_extra_debug.log)

I guess we'd really need to do some more debug in the tryExtractError to see why we are not getting some error back.

We got confirmation from the storage experts that the storage setup that they have does not support posix_fallocate. Its not specifically documented, but some related limitations documented here:
https://www.ibm.com/docs/en/spectrum-scale/5.1.5?topic=management-afm-limitations

**Comment by geoghegk** [comment]
@burqen I also see this issue: https://github.com/neo4j/neo4j/issues/12606#issuecomment-734168120
It talks about falling back to some less efficient emulation if posix_fallocate is not available. Its not clear to me if that is in some other part of the code - but if one part of the code has a fallback plan then all parts should?
maybe @PetrJanouch could comment as he commented on the referenced issue.

**Comment by burqen** [comment]
The last log that you sent is not the debug.log. The docker terminal only display the user log, you'll find the correct log at `logs/debug.log`. You should find the error messages there. Feel free to share it here as well.

The fact that `posix_fallocate` is not supported shouldn't in itself be a problem. Windows and Mac doesn't have support for that and Neo4j run there successfully. So that in itself should be a problem, but your system behave a bit unexpected in other ways as well. In `org.neo4j.kernel.impl.transaction.log.files.TransactionLogChannelAllocator#allocateFile` we only go into `nativeChannelAccessor.preallocateSpace` if file doesn't exist. But in the log you provided it says that the file does exist, `2023-02-24T14:33:02.789919887ZTA-DEBUG /data/transactions/system/neostore.transaction.db.0 DOES EXIST`. You can try and investigate why your jdk `FileSystemProvider` claims that it doesn't exist, because that seems strange.

Another point to investigate is the permissions to the transaction log files. Perhaps those are differently configured than the store files.

I see that you are using OpenJ9 jdk. This is not a supported jdk. You can try and see if changing jdk solves the problem. Supported versions here https://neo4j.com/docs/operations-manual/current/installation/requirements/

I also see that AFM is using NFS (https://www.ibm.com/docs/en/spectrum-scale/5.1.0?topic=management-introduction-active-file-afm), which is not a supported file system. See if you can mount it locally or something to see if that solves the problem. Supported file systems: https://neo4j.com/docs/operations-manual/current/performance/linux-file-system-tuning/.

**Comment by geoghegk** [comment]
> The last log that you sent is not the debug.log.

I attached the debug.log in the earlier message. The last log is showing the cusom logging  from the sys outs that I added. There is nothing new in the last run in the debug.log. We can't reproduce this in house right now because we don't have the same storage setup, and I have limited access to the clients environment, unfortunately I cannot attach the debug.log at this time.

> The fact that posix_fallocate is not supported shouldn't in itself be a problem. Windows and Mac doesn't have support for that and Neo4j run there successfully.

That is not really the point here though - clearly there needs to be a determination as to what OS it is running on, and provide the appropriate native functions in that case. In Neo4J, it looks like it uses the JNA (com.sun.jna, specifically com.sun.jna.Platform) to make that determination. In our case, it clearly determines (correctly) that its running on a Linux env, and that libc is the library with the implementation of the native functions. As far as I can see the method `org.neo4j.internal.nativeimpl.LinuxNativeAccess.tryPreallocateSpace` only has one option, and that is to call `posix_fallocate` . There is no logic there to say, if this fails then try some other function like ftruncate. 

I'm not necessarily suggesting that it should, I was querying if that was the intended behavior. But in any case, as I said earlier, there should be some reporting of this when debug logging is on at least - and I don't see it. Maybe I am not setting the debug logging correctly? I am getting some debug logging for sure - but are there other settings I should be using in addition to the following:
```
NEO4J_dbms_logs_debug_level=DEBUG
```
I also set this one, but wasn't really expecting that to show anything new in this case - which it didn't.
`NEO4J_dbms_logs_query_time__logging__enabled=true`


>In `org.neo4j.kernel.impl.transaction.log.files.TransactionLogChannelAllocator#allocateFile` we only go into `nativeChannelAccessor.preallocateSpace` if file doesn't exist. But in the log you provided it says that the file does exist...

Perhaps I could have done a better job here with the Sys outs that I added and I can see your confusion, but its down to the code. If you look at that method, it first checks if the file exists, and sets a boolean. On initial startup, that boolean will be false. THEN, it opens the file with `fileSystem.write( file );`. After that point, the file does exist (as a 0 sized file) - and my logging correctly reports that it does exist. NOW, there is the check of the boolean which is still false, and so it goes into preallocate. That logic seems fine to me - I don't think there is a problem, except maybe the `fileExist` boolean might be named better, maybe `filePreviouslyExist` - or something.

> Another point to investigate is the permissions to the transaction log files. Perhaps those are differently configured than the store files.

This was one of the first things I looked at I don't see any difference in the permissions.

> I see that you are using OpenJ9 jdk. This is not a supported jdk. 

Yes, We are using OpenJ9 as OpenJDK has a bug in FIPS environments (https://github.com/neo4j/neo4j/issues/12799)
But, I did try OpenJDK, it didn't make any difference - very same behavior.

> I also see that AFM is using NFS...

OK, that may be an issue too. The storage guys tell me that it might be possible to setup a different flavor of AFM that does support posix_fallocate - we will see how far we get with that, but NFS not supported might be an issue too. 

Thanks.

**Comment by burqen** [comment]
Thank you for the extensive reply :)

> I attached the debug.log in the earlier message. ... 

Yes, I see that the first log is the debug.log and the second one is your custom logging. Maybe I'm not being so clear. Let me try again. 

As I understand it, we have two scenarios.
1. Starting dbms on existing store with 0 size tx-log. In this scenario we crash. This is the scenario that is covered by the first debug.log. The theory is that we end up in this scenario because of...
2. Starting dbms the first time, when there are no tx-logs. We will create the file and try to pre-allocate. For you, this fails and it is what is showen by your second log.

Do you share this view?

My theory was that the log messages that you should have seen when we fail to allocate in scenario 2 would be in the debug.log. But they will not be in debug.log you first provided because it doesn't cover that scenario. But you also say that "There is nothing new in the last run in the debug.log", which indeed is very strange.

> That is not really the point here though - clearly there needs to be a determination as to what OS it is running on... As far as I can see the method org.neo4j.internal.nativeimpl.LinuxNativeAccess.tryPreallocateSpace only has one option, and that is to call posix_fallocate . There is no logic there to say, if this fails then try some other function like ftruncate.
>
> I'm not necessarily suggesting that it should, I was querying if that was the intended behavior. But in any case, as I said earlier, there should be some reporting of this when debug logging is on at least - and I don't see it. Maybe I am not setting the debug logging correctly? I am getting some debug logging for sure - but are there other settings I should be using in addition to the following:

Thank you for explaining, it's much clearer for me now. 

I will need to dig a bit deeper to answer if this is intended behaviour. Let me come back to you about this one.

We don't do any diagnostic check about which system calls are available right now, so you are not missing anything there. However we do log the result from the call, which is what I refer to in scenario 2 above. You are setting debug logging correctly.

> Perhaps I could have done a better job here with the Sys outs that I added and I can see your confusion, but its down to the code. If you look at that method, it first checks if the file exists, and sets a boolean. On initial startup, that boolean will be false. THEN, it opens the file with fileSystem.write( file );. After that point, the file does exist (as a 0 sized file) - and my logging correctly reports that it does exist. NOW, there is the check of the boolean which is still false, and so it goes into preallocate. That logic seems fine to me - I don't think there is a problem, except maybe the fileExist boolean might be named better, maybe filePreviouslyExist - or something.

You are absolutely right. I didn't think about the call to fileSystem.write(file).

Another thing that is curious that might be interesting to look into. The error message you get is 28 which is ENOSPC, "there is not enough space left on the device". Since your storage expert claim that the storage doesn't support posix_fallocate I would expect the file system to return EOPNOTSUPP, "The filesystem containing the file referred to by fd does not support this operation.  This error code can be returned by C libraries that don't perform the emulation shown in NOTES, such as musl libc." It might ofc be that the file system just returns the wrong error code, but it looks like an interesting lead.

I see that we do not flush the header after writing it. I will look into doing that as well. Probably only on latest version though.

**Comment by burqen** [comment]
It is intended behaviour that we only rely on posix_fallocate, since our supported file systems are XFS and EXT4 and it is also intended that we don't log if that functionality is missing from underlying system.

This is the git diff that makes it possible to start on top of the empty transaction logs. You can try that out to see if it helps you to start the dbms. You will need to also set `unsupported.dbms.tx_log.fail_on_corrupted_log_files=false`
[patch.txt](https://github.com/neo4j/neo4j/files/10872079/patch.txt)

**Comment by geoghegk** [comment]
Thanks again @burqen 
We finally have a set up in-house to reproduce the issue (i.e. the AFM storage).
I tried out the patch above. Its get a bit further, but eventually fails. This time there is an explicit error in the log that shows the preallocate problem:
```
2023-03-03 16:14:36.758+0000 ERROR [o.n.k.i.t.l.f.LogFileChannelNativeAccessor] [system/00000000] Warning! System is running out of disk space. Failed to preallocate log file since disk does not have enough space left. Please provision more space to avoid that. Allocation failure details: ErrorCode=28, errorMessage='No space left on device'
```
[debug_withpatch.log](https://github.com/neo4j/neo4j/files/10884288/debug_withpatch.log)

I went back and ran it again without the patch. And I think I see the same thing when the container starts up, and the 0 sized transaction log is already there (from the first attempt to start). i.e. I see the explicit error in the log pertaining to the preallocate:

```
2023-03-03 16:21:11.069+0000 ERROR [o.n.k.i.t.l.f.LogFileChannelNativeAccessor] [system/00000000] Warning! System is running out of disk space. Failed to preallocate log file since disk does not have enough space left. Please provision more space to avoid that. Allocation failure details: ErrorCode=28, errorMessage='No space left on device'
```
[debug_nopatch_run2.log](https://github.com/neo4j/neo4j/files/10884287/debug_nopatch_run2.log)

The very first time it is run, when there is no data in /data, we do not see any explicit errors in the log about the preallocate failing, rather we see the failure to read - as was highlighted originally:

```
...Was only able to read -1 bytes, but was expecting 8
```
[debug_nopatch_run1.log](https://github.com/neo4j/neo4j/files/10884287/debug_nopatch_run1.log)

I still feel like I would like to see that explicit message when posix_fallocate fails - and no transaction log file previously existed.

Thanks.

**Comment by burqen** [comment]
@geoghegk I get 404 when trying to open the last log file. Does it work for you? Can you try to upload it again?
